src/heap.cc

```diff
@@ -502,78 +502,81 @@ int ObjectHeap::scavenge() {
            _limit >> KB_LOG2,
            _external_memory >> KB_LOG2);
   }
-  ScavengeState ss(this);
-
-  // Process the roots in the object heap.
-  ss.do_root(reinterpret_cast<Object**>(&_task));
-  ss.do_root(reinterpret_cast<Object**>(&_hatch_arguments));
-  ss.do_roots(_global_variables, program()->global_variables.length());
-  for (auto root : _external_roots) ss.do_roots(root->slot(), 1);
-
-  // Process roots in the _object_notifiers list.
-  for (ObjectNotifier* n : _object_notifiers) n->roots_do(&ss);
-  // Process roots in the _runnable_finalizers.
-  for (FinalizerNode* node : _runnable_finalizers) {
-    node->roots_do(&ss);
-  }
-
-  // Process the to space.
-  Iterator objects(ss.blocks, program());
-  while (!objects.eos()) ss.process_to_objects(objects);
-
-  // Process the registered finalizer list.
-  if (!_registered_finalizers.is_empty() && Flags::tracegc && Flags::verbose) printf(" - Processing registered finalizers\n");
-  ObjectHeap* heap = this;
-  _registered_finalizers.remove_wherever([&ss, heap](FinalizerNode* node) -> bool {
-    bool is_alive = ScavengeState::is_forward_address(node->key()->header_during_gc());
-    if (!is_alive) {
-      // Clear the key so it is not retained.
-      node->set_key(heap->program()->null_object());
-    }
-    node->roots_do(&ss);
-    if (is_alive && Flags::tracegc && Flags::verbose) printf(" - Finalizer %p is alive\n", node);
-    if (is_alive) return false;  // Keep node in list.
-    // From here down, the node is going to be unlinked by returning true.
-    if (Flags::tracegc && Flags::verbose) printf(" - Finalizer %p is unreachable\n", node);
-    heap->_runnable_finalizers.append(node);
-    // Signal finalizers are ready to run.
-    if (heap->_finalizer_notifier != null) {
-      heap->_finalizer_notifier->notify();
-    }
-    return true; // Remove node from list.
-  });
 
-  // Process the finalizers in the to space.
-  while (!objects.eos()) ss.process_to_objects(objects);
-  ASSERT(objects.eos());
+  { ScavengeState ss(this);
```

Comment:
The change is just making sure that the ScavengeState destructor gets called before we compute the amount of memory available after the GC.